test/benchmark/cuda_driver.hpp

```diff
@@ -0,0 +1,220 @@
+/*************************************************************************
+ * Copyright (C) 2018-2022 Blue Brain Project
+ *
+ * This file is part of NMODL distributed under the terms of the GNU
+ * Lesser General Public License. See top-level LICENSE file for details.
+ *************************************************************************/
+
+#pragma once
+
+/**
+ * \dir
+ * \brief Implementation of CUDA and NVVM-based execution engine to run functions from MOD files
+ *
+ * \file
+ * \brief \copybrief nmodl::runner::CUDADriver
+ */
+
+#ifdef NMODL_LLVM_CUDA_BACKEND
+
+#include <memory>
+#include <string>
+
+#include "llvm/IR/Module.h"
+
+#include "benchmark_info.hpp"
+#include "cuda.h"
+#include "cuda_runtime.h"
+#include "gpu_parameters.hpp"
+#include "nvvm.h"
+
+using nmodl::cuda_details::GPUExecutionParameters;
+
+namespace nmodl {
+namespace runner {
+
+struct DeviceInfo {
+    int count;
+    std::string name;
+    int compute_version_major;
+    int compute_version_minor;
+};
+
+/**
+ * @brief Throw meaningful error in case CUDA API call fails
+ *
+ * Checks whether a call to the CUDA API was succsful and if not it throws a runntime_error with
+ * the error message from CUDA.
+ *
+ * @param err Return value of the CUDA API call
+ */
+void checkCudaErrors(CUresult err);
+
+/**
+ * @brief Throw meaningful error in case NVVM API call fails
+ *
+ * Checks whether a call to the NVVM API was succsful and if not it throws a runntime_error with
+ * the error message from NVVM.
+ *
+ * @param err Return value of the NVVM API call
+ */
+void checkNVVMErrors(nvvmResult err);
+
+/**
+ * \class CUDADriver
+ * \brief Driver to execute a MOD file function via the CUDA and NVVM backend.
+ */
+class CUDADriver {
+    /// LLVM IR module to execute.
+    std::unique_ptr<llvm::Module> module;
+    nvvmProgram prog;
+    CUdevice device;
+    CUmodule cudaModule;
+    CUcontext context;
+    CUfunction function;
+    CUlinkState linker;
+    DeviceInfo device_info;
+    std::string ptx_compiled_module;
+
+    void checkCudaErrors(CUresult err);
+    void checkNVVMErrors(nvvmResult err);
+    void load_libraries(BenchmarkInfo* benchmark_info);
+
+  public:
+    explicit CUDADriver(std::unique_ptr<llvm::Module> m)
+        : module(std::move(m)) {}
+
+    /// Initializes the CUDA GPU JIT driver.
+    void init(const std::string& gpu, BenchmarkInfo* benchmark_info = nullptr);
+
+    /// Lookups the entry-point without arguments in the CUDA module and executes it.
+    void execute_without_arguments(const std::string& entry_point,
```

Comment:
`nitpick`: There is a bit of duplication between `execute_without_arguments` and `execute_with_arguments`. The only difference is the line

```c++
// void* kernel_parameters[] = {};
void* kernel_parameters[] = {&arg};
```

Maybe construct `void*` in `run_with_arguments()` and `run_without_arguments()` bodies, and then have a single `execute()` function that  takes `kernel_parameters` as an argument?